VMware.CloudFoundation.CertificateManagement.psm1

```diff
@@ -25,17 +25,900 @@ if ($PSEdition -eq 'Desktop') {
+        $stopwatch = [System.Diagnostics.Stopwatch]::StartNew()
+        do {
+            if ((Test-NetConnection -ComputerName $esxiFqdn -Port 443 -WarningAction SilentlyContinue -ErrorAction SilentlyContinue).TcpTestSucceeded) {
+                if (Connect-VIServer $esxiFqdn -User $user -Password $pass -Force -WarningAction SilentlyContinue -ErrorAction SilentlyContinue) {
+                    $vmHost = Get-VMHost -Server $esxiFqdn
+                    $currentUpTime = New-TimeSpan -Start $vmHost.ExtensionData.Summary.Runtime.BootTime.ToLocalTime() -End (Get-Date)
+                    if ($($esxiUptime.TotalSeconds) -gt $($currentUpTime.TotalSeconds)) {
+                        Write-Host "ESXi host $esxiFqdn has been restarted and is now accessible."
+                    } else {
+                        Write-Host "ESXi host $esxiFqdn uptime: $($esxiUptime.TotalSeconds) | Current Uptime - $($currentUpTime.TotalSeconds)"
+                    }
+                    Disconnect-VIServer -Server $esxiFqdn -Confirm:$false -WarningAction SilentlyContinue -ErrorAction SilentlyContinue | Out-Null
+                    return
+                }
+            }
+            Write-Host "Waiting for ESXi host $esxiFqdn to reboot and become accessible."
+            Start-Sleep -Seconds $pollInterval
+        } while ($stopwatch.elapsed -lt $timeout)
+
+        Write-Error "ESXi host $esxiFqdn did not respond after $($timeout.TotalMinutes) seconds. Please verify that the  ESXi host is online and accessible." -ErrorAction Stop
+    } else {
+        Write-Warning "Restart of ESXi host $esxiFqdn triggered without polling connection state. Please monitor the connection state in the vSphere Client."
+    }        
+}
+
+Export-ModuleMember -Function Restart-EsxiHost
+
+
+Function Install-EsxiCertificate {
+    <#
+        .SYNOPSIS
+        Install a certificate for an ESXi host or for each ESXi host in a cluster.
+
+        .DESCRIPTION
+        The Install-EsxiCertificate cmdlet will replace the certificate for an ESXi host or for each ESXi host in a cluster
+        (the behavior is controlled with parameter -cluster/-esxiFqdn). 
+        You must provide the directory containing the signed certificate files
+        Certificate names should be in format <FQDN>.crt e.g. sfo01-m01-esx01.sfo.rainpole.io.crt
+        The workflow will put ESXi host in maintenance mode with full data migration, 
+        disconnect ESXi host from the vCenter Server, replace the certificate, reboot the ESXi host, and exits maintenance mode once the ESXi host is online.
+        Timeout for putting ESXi host in maintenance is provided in seconds using -timeout Parameter. Default is 18000 seconds or 5 hrs. 
+
+        .EXAMPLE
+        Install-EsxiCertificate -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -esxiFqdn sfo01-m01-esx03.sfo.rainpole.io -certificateDirectory F:\certificates -certificateFileExt ".cer"
```

Comment:
```suggestion
        Install-EsxiCertificate -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-m01 -esxiFqdn sfo01-m01-esx01.sfo.rainpole.io -certificateDirectory F:\certificates -certificateFileExt ".cer"
```

Reply:
Based on the parameters it defaults to .cer so it's not required in the example.

Reply:
I think it is better to keep it because user may not have the file in .cer ext.

Reply:
Works for me. Consider one example with and one without.

Reply:
Changed to ValidateSet

Reply:
Discussed. Closing this.